packages/sitecore-jss/src/graphql/search-service.ts

```diff
@@ -0,0 +1,121 @@
+import { GraphQLClient } from '../graphql-request-client';
+
+/**
+ * Schema of data returned in response to a "search" query request
+ * @template T The type of objects being requested.
+ */
+export type SearchQueryResult<T> = {
+  search: {
+    /**
+     * Data needed to paginate the search results
+     */
+    pageInfo: {
+      /**
+       * string token that can be used to fetch the next page of results
+       */
+      endCursor: string;
+      /**
+       * a value that indicates whether more pages of results are available
+       */
+      hasNext: boolean;
+    };
+    /*
+     * the type of data querying about items matching the search criteria
+     */
+    results: T[];
+  };
+};
+
+/**
+ * Describes the variables used by the 'search' query. Language should always be specified.
+ * The other predicates are optional.
+ */
+export type SearchQueryVariables = {
+  /**
+   * Required. The language versions to search for. Fetch pages that have versions in this language.
+   */
+  language: string;
+
+  /**
+   * Optional. The ID of the search root item. Fetch items that have this item as an ancestor.
+   */
+  rootItemId?: string;
+
+  /**
+   * Optional. How many result items to fetch in each GraphQL call. This is needed for pagination.
+   * @default 10
+   */
+  pageSize?: number;
+
+  /**
+   * Optional. Sitecore template ID(s). Fetch items that inherit from this template(s).
+   */
+  templates?: string;
+};
+
+/**
+ * Configuration options for service classes that extend @see SearchQueryService.
+ * This extends @see SearchQueryVariables because properties that can be passed to the search query
+ * as predicates should be configurable. 'language' is excluded because, normally, all properties
+ * except 'language' are consistent across languages so they are passed to constructors, and
+ * 'language' can vary so it is passed to methods.
+ */
+export interface SearchServiceConfig extends Omit<SearchQueryVariables, 'language'> {
```

Comment:
The comment helps, but I still found this a little confusing (interaction between `SearchQueryVariables` and `SearchServiceConfig` and their usages).